File: leetcode/3152-special-array-ii/SpecialArrayII.java
import java.util.Arrays;

public class SpecialArrayII {
  public static void main(String[] args) {
    var numbers = new int[] { 3, 4, 1, 2, 6 };
    var queries = new int[][] { { 0, 4 } };

    System.out.println(Arrays.toString(new Solution().isArraySpecial(numbers, queries)));
  }
}

class Solution {
  public boolean[] isArraySpecial(int[] numbers, int[][] queries) {
    int[] specials = new int[numbers.length];

    for (int i = 1; i < specials.length; i++) {
      if (numbers[i] % 2 != numbers[i - 1] % 2) {
        specials[i] = specials[i - 1];

        continue;
      }

      specials[i] = i;
    }

    boolean[] results = new boolean[queries.length];

    for (int i = 0; i < results.length; i++) {
      results[i] = specials[queries[i][1]] <= specials[queries[i][0]];
    }

    return results;
  }
}

File: leetcode/3001-4000/3101-3200/3152-special-array-ii/SpecialArrayII.ts
function isArraySpecial(nums: number[], queries: number[][]): boolean[] {
  const specials: number[] = new Array(nums.length);

  specials[0] = 0;

  for (let i = 1; i < nums.length; i++) {
    if (nums[i] % 2 !== nums[i - 1] % 2) {
      specials[i] = specials[i - 1];
      continue;
    }

    specials[i] = i;

  }

  return queries.map(q => specials[q[1]] <= specials[q[0]]);
}


// Testcases
const nums = [3, 4, 1, 2, 6];
const queries = [[0, 1]];
// const queries = [[0, 4]];

console.log(isArraySpecial(nums, queries));
